drivers/mtd/gd55.c

```diff
@@ -0,0 +1,2065 @@
+ *
+ *   MTD devices are not registered in the file system, but are created as
+ *   instances that can be bound to other functions (such as a block or
+ *   character driver front end).
+ *
+ * Input Parameters:
+ *   qspi       - a reference to the qspi device to initialize
+ *   unprotect  - if true, unprotect the device
+ *
+ * Returned Value:
+ *   Success (OK) or fail (negated error code)
+ ****************************************************************************/
+
+FAR struct mtd_dev_s *gd55_initialize(FAR struct qspi_dev_s *qspi,
+                                         bool unprotect)
+{
+  FAR struct gd55_dev_s *dev;
+  int ret;
+  uint8_t status;
+  uint16_t config;
+
+  DEBUGASSERT(qspi != NULL);
+
+  /* Allocate a state structure (we allocate the structure instead of using
+   * a fixed, static allocation so that we can handle multiple FLASH devices.
+   * The current implementation would handle only one FLASH part per QuadSPI
+   * device (only because of the QSPIDEV_FLASH(0) definition) and so would
+   * have to be extended to handle multiple FLASH parts on the same QuadSPI
+   * bus.
+   */
+
+  dev = kmm_zalloc(sizeof(*dev));
+
+  if (dev == NULL)
+    {
+      ferr("Failed to allocate mtd device\n");
+      return NULL;
+    }
+
+  dev->mtd.erase  = gd55_erase;
+  dev->mtd.bread  = gd55_bread;
+  dev->mtd.bwrite = gd55_bwrite;
+  dev->mtd.read   = gd55_read;
+  dev->mtd.ioctl  = gd55_ioctl;
+  dev->mtd.name   = "gd55";
+  dev->qspi       = qspi;
+
+  /* Allocate a 4-byte buffer to support DMA-able command data */
+
+  dev->cmdbuf = (FAR uint8_t *)QSPI_ALLOC(qspi, 4);
+  if (dev->cmdbuf == NULL)
+    {
+      ferr("Failed to allocate command buffer\n");
+      goto exit_free_dev;
+    }
+
+  dev->readbuf = (FAR uint8_t *)QSPI_ALLOC(qspi, 2);
+  if (dev->readbuf == NULL)
+    {
+      ferr("ERROR Failed to allocate read buffer\n");
+      goto exit_free_cmdbuf;
+    }
+
+  /* Identify the FLASH chip and get its capacity */
+
+  ret = gd55_readid(dev);
+  if (ret != OK)
+    {
+      /* Unrecognized! Discard all of that work we just did and return NULL */
+
+      ferr("Unrecognized QSPI device\n");
+      goto exit_free_readbuf;
+    }
+
+  /* Unprotect all FLASH sectors if so requested. */
+
+  if (unprotect)
+    {
+      ret = gd55_unprotect(dev, 0, dev->nsectors - 1);
+      if (ret < 0)
+        {
+          ferr("ERROR: Sector unprotect failed\n");
+        }
+    }
+
+#ifdef CONFIG_MTD_GD55_SECTOR512  /* Simulate a 512 byte sector */
+  /* Allocate a buffer for the erase block cache */
+
+  dev->sector = (FAR uint8_t *)QSPI_ALLOC(qspi, GD55_SECTOR_SIZE);
+  if (dev->sector == NULL)
+    {
+      /* Allocation failed! Discard all of that work we just did and
+       * return NULL
+       */
+
+      ferr("ERROR: Sector allocation failed\n");
+      goto exit_free_readbuf;
+    }
+#endif
+
+  /* Avoid compiler warnings in case info logs are disabled */
+
+  UNUSED(status);
```

Comment:
remove unnecessary code

Reply:
The perils of copying preexisting code!! I will check whether there ought to be some debug type code here using those and amend as appropriate. Thanks.

Reply:
done